### ISSUE TITLE
htlcswitch+rpcserver: add new streaming ForwardingEvent API

### ISSUE THREAD

**Issue by Roasbeef** [lightningnetwork/lnd#3420]
Today users are able to poll the `ForwardingHistory` API in to obtain the past history of successful HTLC forwards. This is useful to be able to gauge if your node is forwarding payments at all, but only exposing successful payment settles doesn't provide enough information to be able to make channel management/allocation decisions. The original API didn't also move to _persist_ HTLC failures as they can be much more numerous than failures since they have low cost today in the network. However, we can still surface this information based on our in-memory state, and rely on projects such as `lndmon` to persist this information for analysis purposes. 

In order to provide node operators with additional information that they can use to optimize the channel balances and placement, we should expose a _streaming_ variant of the `ForwardingHistory` API that also includes HTLCs accepts _and_ fails. The final API should be able to export events pertaining to: successful HTLC forwards, multi-hop failures, failed forwarding attempts and successful payment settlement.

# Steps To Completion

- [ ] Extend the [`htlcswitch`](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/switch.go#L176) to expose a subscription API to allow clients to be notified of relevant HTLC forwarding events. 

- [ ] A sample draft of relevant forwarding events may resemble something like the following:
```go
type ForwardingEvent struct {
    IncomingChannel, OutgoingChannel lnwire.ShortChannelID

    IncomingTimeLock, OutgoingTimeLock uint32
    
    IncomingAmt, OutgoingAmt lnwire.Milisatoshi

    PayHash [32]byte
}

type LinkFailEvent struct { // only for HTLCS that come in, but fail outgoing link requirements (capacity, fees, etc)
    FailReason FailReason // new enum exposing various failure reasons
 
    ForwardingEvent
}

type ForwardingFailEvent  struct { // no error as multi-hop errors are opaque
    ElapsedTime time.Duration // amount of time it took for us to recv the multi-hop fail

    ForwardingEvent 
}

type SettleEvent struct {
    ElapsedTime time.Duration 

    ForwardingEvent
}
```

- [ ] Create a new `SubscribeForwardingEvents` streaming API that maps something similar to the structs above to actual protos. Note that since we don't persist the information in this form, we won't support any sort of backlog filling.

**Comment by carlaKC** [comment]
I've started having a look at implementing this, starting off looking at the code in `switch.go#handlePacketForward`. 

A few questions that have come up:
1. Do we want to include local (originating from our node) events in this stream? I think they will be relevant to `channelinsights`?

2. We have link failures in a few cases, do we want to report all of them? I would imagine yes, since it's giving us a full picture of what's happening in the link, but wanted to double check.
   - We're rejecting all HTLCs
   - We can't find the link
   - We don't have a link with sufficient capacity
   - Policy reasons

**Comment by Roasbeef** [comment]
1. We already have that information in the database for `ListPayments` as we track the full route used. 
2. I think reporting the _reason_ is a nice to have beyond reporting that a failure occurred in the first place. I agree it can be useful when making optimization decision (failed due to invalid policy vs failed due to insufficient bandwidth). I think the base set of reasons you enumerated is a good set to start off with, as they all seem actionable from the PoV of a channel management optimizer.

**Comment by nicolasburtey** [comment]
This would be useful to have some interactive validation of the HTLC forwarding, versus simply reporting like this PR is planning on.

Think of a use case where a routing Alice is connecting to a mobile phone Bob, knowing Bob is only coming online once a day or so. Alice will want to hold the HTLC for one day until Bob come online. 

Having a callback on a potential forward HTLC would enable that use case. It seems, based on a discussion with @alexbosworth today that currently, Alice would automatically fail if Bob is not online.

**Comment by carlaKC** [comment]
> Having a callback on a potential forward HTLC would enable that use case. It seems, based on a discussion with @alexbosworth today that currently, Alice would automatically fail if Bob is not online.

A callback for htlc acceptance has some interesting applications for controlling your node's forwarding behavior beyond what is available in the protocol, but I it would be a separate change in the code (pretty independent of the current PR). If this is a feature that you'd like in lnd, please do open an issue (or even better a PR I'd be happy to review)!

**Comment by nicolasburtey** [comment]
ok, I'll create an issue for now!

> A callback for htlc acceptance has some interesting applications for controlling your node's forwarding behavior beyond what is available in the protocol

just to clarify, by protocol you mean lnd? or lightning in general?

**Comment by carlaKC** [comment]
The issue would be in lnd, since this change would be on an implementation level.

> just to clarify, by protocol you mean lnd? or lightning in general?

Currently lnd lets you control forwards based on the control allowed by the protocol (max htlc size/fees etc). A htlc acceptor would add additional control which is more customizable than the protocol.

**Comment by MrManPew** [comment]
Will this also tell you if the payment was part of a AMP/MMP payment?

**Comment by carlaKC** [comment]
> Will this also tell you if the payment was part of a AMP/MMP payment?

It will not, we notify htlcs on the switch level, which is ignorant of mpp, all it knows is the forwarding/settling/failing of htlcs.

@MrManPew are you looking to detect MPP payments which your node forwards for other people, or MPP payment initiated by your own node? You can detect that your own sends are MPP payments by looking for a MPP record in `listpayments`:
```
"mpp_record": {
    "payment_addr": "756521bc2b78d...9cbe57f3d55bab82adc7839ca12",
    "total_amt_msat": "1000"
},
```

**Comment by MrManPew** [comment]
Thanks @carlaKC , I am actually interested in MPP payment which my node forwards. Is this information not available to the node, or is it just not captured?

**Comment by carlaKC** [comment]
> Is this information not available to the node, or is it just not captured?

This information isn't strictly available to you as a node. You may observe multiple payments being forwarded through your node with the same payment hash, and chances are that they're _probably_ a MPP payment. But since the rest of the route is onion encrypted, you don't know who the sender/receiver is, so you can't tell whether they're part of a MPP payment, or somebody is simply reusing an invoice.

**Comment by MrManPew** [comment]
Great thanks

**Comment by MrManPew** [comment]
Hi @Roasbeef  and @carlaKC, maybe I am missing it, but I have not seen anything in[ v0.10 release log](https://github.com/lightningnetwork/lnd/releases/tag/v0.10.0-beta.rc4) about this feature...

Thanks

**Comment by MrManPew** [comment]
Hi @Roasbeef and @carlaKC , just to be sure, there is no mention on this feature in the release log yet - https://github.com/lightningnetwork/lnd/releases/tag/v0.10.0-beta.rc5
Correct ?

**Comment by carlaKC** [comment]
Hey @MrManPew, the release notes are still a work in progress (and I was a bit late on writing mine up 😬). The feature has been merged, and will be in lnd v10.0. 

We're always looking for people to test the release candidates, so if you'd like to check out rc5 and test out the stream that would be great!

**Comment by MrManPew** [comment]
Thank @carlaKC ,I'll be happy to help with the testing, but I don't know how to use this new feature (as there are no instructions available :) )
